File: src/events/component_interaction/kind_confirm.rs
use poise::serenity_prelude as serenity;

use serenity::*;

use crate::prelude::*;

pub async fn yes_clicked(
    context: &Context,
    component_interaction: &MessageComponentInteraction,
    bot_data: &BotData,
    name: String,
    kind: VerifiedUserKind,
) -> BotResult<()> {
    let user = &component_interaction.user;

    // Confirm the user as verified, but not registered yet
    sqlx::query!(
        "update registrations
        set status = ?, name = null, kind = null
        where user_id = ?",
        RegistrationStatus::Verified.to_string(),
        user.id.0,
    )
    .execute(&bot_data.database_pool)
    .await?;

    // Link the user's account to the verified user
    sqlx::query!(
        "update users
        set registered_user_id = ?
        where name = ? and kind = ? and registered_user_id is null
        limit 1", // NOTE: Only one user should be consdiered verified
        user.id.0,
        name,
        kind.to_string(),
    )
    .execute(&bot_data.database_pool)
    .await?;

    // Inform the user of verification and extra questions
    component_interaction
        .create_followup_message(&context.http, |message| {
            message
                .embed(embeds::registration::verified())
                .components(components::verified_continue_button())
        })
        .await?;

    // Log the verification
    bot_data
        .channels
        .log_channel_id
        .send_message(&context.http, |message| {
            message.embed(embeds::logs::verified(user.id, kind))
        })
        .await?;

    Ok(())
}

pub async fn no_clicked(
    context: &Context,
    component_interaction: &MessageComponentInteraction,
    bot_data: &BotData,
    kind: VerifiedUserKind,
) -> BotResult<()> {
    let user = &component_interaction.user;

    // Set the user's registration state to failed
    sqlx::query!(
        "update registrations
        set status = ?, name = null
        where user_id = ?",
        RegistrationStatus::Failed.to_string(),
        user.id.0,
    )
    .execute(&bot_data.database_pool)
    .await?;

    // Ask the user to seek assistance
    component_interaction
        .create_followup_message(&context.http, |message| {
            message.embed(embeds::registration::kind_error(
                bot_data.channels.support_channel_id,
            ))
        })
        .await?;

    // Alert the mentors about incorrect kind
    bot_data
        .channels
        .log_channel_id
        .send_message(&context.http, |message| {
            message
                .content(format!("<@&{}>", bot_data.roles.hierarchy.mentor_role_id))
                .embed(embeds::logs::kind_error(
                    user.id,
                    kind,
                    bot_data.channels.support_channel_id,
                ))
        })
        .await?;

    Ok(())
}

File: norris/events/component_interaction/kind_confirm.py
from discord import Interaction
from sqlalchemy import select, update

from ...bot import Norris
from ...model import KindFound, Registration, RegistrationStatus, VerifiedUser
from . import verify_registration_status


async def yes_clicked(interaction: Interaction, norris: Norris) -> None:
    # Defer response to give time for database queries
    await interaction.response.defer()

    # Check that the user has the correct state to click the button
    if not await verify_registration_status(interaction.user.id,
                                            RegistrationStatus.KIND_FOUND,
                                            norris):
        return

    async with norris.database_engine.begin() as connection:
        # Get the user's name and kind
        result = await connection.execute(
            select(KindFound)
            .where(Registration.user_id == interaction.user.id)
            .limit(1),
        )
        registration = result.one()

        # Update the user's registration state to verified
        await connection.execute(
            update(Registration)
            .values(status=RegistrationStatus.VERIFIED, name=None, kind=None)
            .where(Registration.user_id == interaction.user.id),
        )

        # Select the ID of a verified user with a matching name and role
        # NOTE: we need this because sqlalchemy doesn't support limits on updates, which
        # means we can't just limit the change to one row like in the Rust version
        result = await connection.execute(
            select(VerifiedUser)
            .where(VerifiedUser.name == registration.name
                   and VerifiedUser.kind == registration.kind
                   and VerifiedUser.registered_user_id is None)
            .limit(1),
        )
        verified_user = result.one()

        # Link the user's account to the matching verified user
        await connection.execute(
            update(VerifiedUser)
            .values(registered_user_id=interaction.user.id)
            .where(VerifiedUser.id == verified_user.id),
        )

    # NOTE: thanks Python for this amazing module system
    from ...responses import embeds
    from ...responses.components import VerifiedContinueView

    # Inform the user of verification and ask them to continue with optional questions
    await interaction.followup.send(
        embed=embeds.registration.verified(),
        view=VerifiedContinueView(norris),
    )

    # Log the verification
    await norris.get_channel(norris.channels.log_channel_id).send(
        embed=embeds.logs.verified(interaction.user.id, verified_user.kind),
    )


async def no_clicked(interaction: Interaction, norris: Norris) -> None:
    # Defer response to give time for database queries
    await interaction.response.defer()

    # Check that the user has the correct state to click the button
    if not await verify_registration_status(interaction.user.id,
                                            RegistrationStatus.KIND_FOUND,
                                            norris):
        return

    async with norris.database_engine.begin() as connection:
        # Get the user's name and kind
        result = await connection.execute(
            select(KindFound)
            .where(Registration.user_id == interaction.user.id)
            .limit(1),
        )
        registration = result.one()

    # Update the user's registration state to failed
    async with norris.database_engine.begin() as connection:
        await connection.execute(
            update(Registration)
            .values(status=RegistrationStatus.FAILED, name=None, kind=None)
            .where(Registration.user_id == interaction.user.id),
        )

    # NOTE: I want to bang my head against a wall
    from ...responses import embeds

    # Ask the user to seek assistance
    await interaction.followup.send(
        embed=embeds.registration.kind_error(norris.channels.support_channel_id),
    )

    # Alert mentors about kind error
    await norris.get_channel(norris.channels.log_channel_id).send(
        content=f"<@&{norris.roles.hierarchy.mentor_role_id}>",
        embed=embeds.logs.kind_error(interaction.user.id,
                                     registration.kind,
                                     norris.channels.support_channel_id),
    )
